### ISSUE TITLE
wasm: optimize proxy_wasm::null_plugin::getProperty and/or usages of it

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#33295]
![2021-06-06_18-37-30](https://user-images.githubusercontent.com/623453/120948086-419d1480-c6f6-11eb-99ee-3de49c79fcfc.png)

Above shows pprof on inbound HTTP proxy, filtered on getProperty. 10% of CPU is spent on getProperty, which seems potentially excessive.

It seems like there are some low hanging fruit here, if I understand the code correctly (which is a big if, this is my first time looking at it). getTrafficDirection (0.6%) and getServiceNameFallback (1.34%) are per-listener and per-proxy settings(?), but are looked up on each stream.

populateRequestInfo (5.58%, including the 1.34 from getServiceNameFallback) also seems to have a fair amount of somewhat static values. cluster_name and route_name seem like they are essentially static, and there are a few connection level ones which could possibly be looked up per connection instead of per request.

We may also get benefits from just making getProperty faster, if feasible. It seems to be doing some pretty heavy protobuf stuff. Perhaps we can fetch multiple values at once to amortize the cost? I have no clue how feasible this is

**Comment by bianpengyuan** [comment]
This is a known side effect of using Wasm API, which performs unnecessary copies. Tencent rewrote this filter to native and gain ~15%-20% cpu savings:

<img width="1772" alt="Screen Shot 2021-06-06 at 8 50 40 PM" src="https://user-images.githubusercontent.com/34738376/120956743-12dc6980-c709-11eb-8f4e-15c437d25b9d.png">

When it becomes important, we can revisit whether it is still desired to keep stats filter as a Wasm filter, since it is unlikely we will ever run it as Wasm.

**Comment by howardjohn** [comment]
I get there is some overheads of copying, but it seems they are beyond what is required. I am not an expert though. Can we not copy everything in one operation instead of 20+ copies? And can we extract constant values once instead of on every request (may not be required if we do the previous optimization)?

**Comment by bianpengyuan** [comment]
I think we've already avoided copying repeatedly per request and small copies, when it is easy :). Not sure how much we'd want to invest into harder optimization like copy per connection instead of per request (which is very tricky and error prone I think), versus a refactor.

**Comment by mathetake** [comment]
I see that this seems not stemming from copying attributes (which is plausible as the value would be in a few bytes in anyway) - looks like most of the time is spent in findValue and cel-cpp, especially when the path involves with map types in there.

Maybe we could do some caching or treating the "hot" property path specially in Envoy implementation (Context::findValue and Context::getProperty).

**Comment by howardjohn** [comment]
@mathetake that is exactly what I was thinking. Can we somehow amortize the cost? Right now we are calling the full findValue stuff ~10 times per request. if we are able to do one request that fetches everything we need I would think that could be substantially more performant?

**Comment by howardjohn** [comment]
Not stale

**Comment by huanghuangzym** [comment]
> This is a known side effect of using Wasm API, which performs unnecessary copies. Tencent rewrote this filter to native and gain ~15%-20% cpu savings:
> 
> <img alt="Screen Shot 2021-06-06 at 8 50 40 PM" width="1772" src="https://user-images.githubusercontent.com/34738376/120956743-12dc6980-c709-11eb-8f4e-15c437d25b9d.png">
> 
> When it becomes important, we can revisit whether it is still desired to keep stats filter as a Wasm filter, since it is unlikely we will ever run it as Wasm.

hello ,can help to show how to rewrite stats filter to gain cpu use?

**Comment by zirain** [comment]
what he said reference from https://cloud.tencent.com/developer/article/1685873, but it's not open source.

**Comment by huanghuangzym** [comment]
> what he said reference from https://cloud.tencent.com/developer/article/1685873, but it's not open source.

It's a pity, would you consider open source?

**Comment by zirain** [comment]
> > what he said reference from https://cloud.tencent.com/developer/article/1685873, but it's not open source.
> 
> It's a pity, would you consider open source?

IIRC, Tencent saild there's no plan for open source `tstats` in that talk.
I have no updated for it, maybe you should ask someone in `aeraki-mesh`?

**Comment by howardjohn** [comment]
I believe (but please verify before working on it...) we would be open to a rewrite of the stats system to not use WASM. We had someone working on it but unfortunately it fell through the cracks

**Comment by huanghuangzym** [comment]
> I believe (but please verify before working on it...) we would be open to a rewrite of the stats system to not use WASM. We had someone working on it but unfortunately it fell through the cracks

i would like to have a try

**Comment by kyessenov** [comment]
We did not actually attempt to rewrite the stats filter from Wasm. We focused on our vendor-specific stackdriver extension (which did not improve things much because ultimately it runs gRPC callouts). You are welcome to submit a PR to istio/proxy with a rewrite of stats filter into a native extension. There is a variety of low hanging fruit that will easily make it perform better:
* remove Wasm copying overhead;
* improve stats lifecycle management (expiry);
* remove regex tag extraction overhead;
* remove protobuf/base64 CPU overhead by using simpler encoding;
* remove flatbuffers, they are only needed because Wasm.

Given all the above, it's actually surprising stats performs as well as it does, and you only get 20% improvement. I think it's probably more than that if you dig deeper.

Our stats implementation was a demonstration of Wasm capability, and it's still the best Wasm extension we have. So it took a lot of overhead to make it happen, and the performance was only secondary priority.
